new/default-resource-name.md

```diff
@@ -0,0 +1,103 @@
+---
+RFC: unassigned
+Author: John Keiser <jkeiser@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Default Resource Name
+
+In which resources get names and DSL by default, unless you say you don't want one.
+
+## Motivation
+
+    As a Chef user,
+    I want convention to drive my DSL,
+    So that I can concentrate on my code instead of ceremony.
+
+## Specification
+
+When a user declares a resource class thus:
+
+```ruby
+class MyResource < Chef::Resource
+end
+```
+
+It will automatically gain a resource_name and DSL derived from the class:
+
+```ruby
+my_resource 'foo' do
+end
+```
+
+This is already (currently) the case for classes declared in Chef::Resource; this extends the rule to all classes anywhere and removes the special case.
+
+### resource_name creates automatic DSL
+
+When a class's resource_name is set, it will always get DSL will that resource name. Thus, this will work:
+
+```ruby
+class X < Chef::Resource
+  resource_name :my_resource
+end
+my_resource 'foo' do
+end
+```
+
+#### New resource classes' resource names are set automatically
+
+When a resource class is declared with an explicit name, its resource name is set automatically:
+
+```ruby
+class SuperAwesome < Chef::Resource
```

Comment:
@coderanger I'm in full concurrence with what you are talking about, I think.  I could expand this proposal to say something like:

`Chef::Resource` classes named `X::Y::Z` or `X::Y::Z::Resource` get automatic resource name `z`.  Additionally, if no provider is automatically found for the resource, we will look for a `Chef::Provider` class named `X::Y::Z::Provider`, and use that.

`Chef::Provider::X` and `Chef::Resource::X` will *not* be considered pairs in this hierarchy, and `Chef::Provider::X` will *not* be given an automatic `provides` line.  In fact, no provider will *ever* be given an automatic `provides` line; pairing is the only way it will be automatically chosen.

Core Chef providers in `Chef::Provider` will move to `Chef::Resource::X::Provider` as well, and alised into the `Chef::Provider` namespace for backwards compatibility.

The intent here is that Providers should be (by default) rather tightly coupled with Resources (see @lamont-granquist's dreams of 1:1 mapping). In general, unless you actually create multiple providers which match up with the exact same resource class, you should never ever call `provides` on a provider.